File: exec/javascript/src/builtin/fs.js
import {
    read_file,
    read_file_with_encoding,
    write_file,
    write_file_with_encoding
} from '__golem_exec_js_builtin/fs_native';
import {Buffer} from 'node:buffer';

export function readFile(path, optionsOrCallback, callback) {
    if (typeof optionsOrCallback === 'function') {
        callback = optionsOrCallback;
        optionsOrCallback = {};
    }
    if (typeof optionsOrCallback === 'string') {
        optionsOrCallback = {encoding: optionsOrCallback};
    }
    if (optionsOrCallback.encoding && optionsOrCallback.encoding !== '') {
        const [contents, error] = read_file_with_encoding(path, optionsOrCallback.encoding);
        if (error === undefined) {
            callback(contents);
        } else {
            callback(undefined, error);
        }
    } else {
        const [contents, error] = read_file(path);
        if (error === undefined) {
            const buffer = Buffer.from(contents);
            callback(buffer);
        } else {
            callback(undefined, error);
        }
    }
}

export function readFileSync(path, options) {
    if (typeof options === 'string') {
        options = {encoding: options};
    }
    if (options && options.encoding && options.encoding !== '') {
        const [contents, error] = read_file_with_encoding(path, options.encoding);
        if (error === undefined) {
            return contents;
        } else {
            throw new Error(error);
        }
    } else {
        const [contents, error] = read_file(path);
        if (error === undefined) {
            return Buffer.from(contents);
        } else {
            throw new Error(error);
        }
    }
}

export function writeFile(path, data, optionsOrCallback, callback) {
    if (typeof optionsOrCallback === 'function') {
        callback = optionsOrCallback;
        optionsOrCallback = {};
    }
    if (typeof optionsOrCallback === 'string') {
        optionsOrCallback = {encoding: optionsOrCallback};
    }
    if (optionsOrCallback && optionsOrCallback.encoding && optionsOrCallback.encoding !== '') {
        const error = write_file_with_encoding(path, optionsOrCallback.encoding, data);
        callback(error);
    } else {
        if (typeof data === 'string') {
            const error = write_file_with_encoding(path, "utf8", data);
            callback(error);
        } else {
            const dataArray = new Uint8Array(data.buffer || data, data.byteOffset || 0, data.byteLength || data.length);
            const error = write_file(path, dataArray);
            callback(error);
        }
    }
}


export function writeFileSync(path, data, options) {
    if (typeof options === 'string') {
        options = {encoding: options};
    }
    if (options && options.encoding && options.encoding !== '') {
        const error = write_file_with_encoding(path, options.encoding, data);
        if (error !== undefined) {
            throw new Error(error);
        }
    } else {
        if (typeof data === 'string') {
            const error = write_file_with_encoding(path, "utf8", data);
            if (error !== undefined) {
                throw new Error(error);
            }
        } else {
            const dataArray = new Uint8Array(data.buffer || data, data.byteOffset || 0, data.byteLength || data.length);
            const error = write_file(path, dataArray);
            if (error !== undefined) {
                throw new Error(error);
            }
        }
    }
}

File: exec/javascript/src/builtin/fs.rs
use golem_exec::golem::exec::types::Error;
use rquickjs::{Ctx, JsLifetime};
use std::path::{Path, PathBuf};

#[rquickjs::module(rename_vars = "camelCase")]
pub mod native_module {
    use encoding_rs::Encoding;
    use rquickjs::prelude::List;
    use rquickjs::{Ctx, TypedArray};

    #[rquickjs::function]
    pub fn read_file_with_encoding(
        path: String,
        encoding: String,
        ctx: Ctx<'_>,
    ) -> List<(Option<String>, Option<String>)> {
        let path = super::resolve_path(&ctx, &path);
        match std::fs::read(&path) {
            Ok(bytes) => match Encoding::for_label(encoding.as_bytes()) {
                Some(encoding) => {
                    let (decoded, _) = encoding.decode_with_bom_removal(&bytes);
                    let decoded_string = decoded.into_owned();
                    List((Some(decoded_string), None))
                }
                None => List((None, Some(format!("Unsupported encoding: {}", encoding)))),
            },
            Err(err) => {
                let error_message = format!("Failed to read file {path:?}: {err}");
                List((None, Some(error_message)))
            }
        }
    }

    #[rquickjs::function]
    pub fn read_file(
        path: String,
        ctx: Ctx<'_>,
    ) -> List<(Option<TypedArray<'_, u8>>, Option<String>)> {
        let path = super::resolve_path(&ctx, &path);
        match std::fs::read(&path) {
            Ok(bytes) => {
                let typed_array =
                    TypedArray::new_copy(ctx.clone(), &bytes).expect("Failed to create TypedArray");
                List((Some(typed_array), None))
            }
            Err(err) => {
                let error_message = format!("Failed to read file {path:?}: {err}");
                List((None, Some(error_message)))
            }
        }
    }

    #[rquickjs::function]
    pub fn write_file_with_encoding(
        path: String,
        encoding: String,
        content: String,
        ctx: Ctx<'_>,
    ) -> Option<String> {
        if encoding != "utf8" {
            Some("Only 'utf8' encoding is supported".to_string())
        } else {
            let bytes = content.as_bytes();
            let path = super::resolve_path(&ctx, &path);
            if let Some(parent) = path.parent() {
                if let Err(err) = std::fs::create_dir_all(parent) {
                    return Some(format!(
                        "Failed to create directory {}: {}",
                        parent.display(),
                        err
                    ));
                }
            }
            if let Err(err) = std::fs::write(&path, bytes) {
                Some(format!("Failed to write file {path:?}: {err}"))
            } else {
                None // Success
            }
        }
    }

    #[rquickjs::function]
    pub fn write_file(path: String, content: TypedArray<'_, u8>, ctx: Ctx<'_>) -> Option<String> {
        if let Some(bytes) = content.as_bytes() {
            let path = super::resolve_path(&ctx, &path);
            if let Some(parent) = path.parent() {
                if let Err(err) = std::fs::create_dir_all(parent) {
                    return Some(format!(
                        "Failed to create directory {}: {}",
                        parent.display(),
                        err
                    ));
                }
            }
            if let Err(err) = std::fs::write(&path, bytes) {
                Some(format!("Failed to write file {path:?}: {err}"))
            } else {
                None // Success
            }
        } else {
            Some("The typed array has been detached".to_string())
        }
    }
}

#[derive(Debug, Clone, JsLifetime)]
struct FsConfig {
    pub data_root: PathBuf,
}

fn get_data_root(ctx: &Ctx<'_>) -> PathBuf {
    let fs_config = ctx.userdata::<FsConfig>().unwrap();
    fs_config.data_root.clone()
}

fn resolve_path(ctx: &Ctx<'_>, path: &str) -> PathBuf {
    let data_root = get_data_root(ctx);
    let resolved_path = if path.starts_with('/') {
        data_root.join(&path[1..])
    } else {
        let cwd: String = ctx
            .globals()
            .get("__golem_exec_js_cwd")
            .expect("Failed to get cwd");
        let cwd = cwd.trim_start_matches('/');
        data_root.join(cwd).join(path)
    };

    if !resolved_path.starts_with(&data_root) {
        panic!("Path {resolved_path:?} is outside the data root {data_root:?}",);
    }

    resolved_path
}

pub fn init_fs(ctx: Ctx<'_>, data_root: &Path) -> Result<(), Error> {
    let fs_config = FsConfig {
        data_root: data_root.to_path_buf(),
    };
    ctx.store_userdata(fs_config)
        .map_err(|err| Error::Internal(err.to_string()))?;
    Ok(())
}

// JS functions for the fs implementation
pub const FS_JS: &str = include_str!("fs.js");
